app/code/Magento/InventoryShipping/Observer/VirtualSourceDeductionProcessor.php

```diff
@@ -0,0 +1,271 @@
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->inventoryRequestFactory = $inventoryRequestFactory;
+        $this->itemRequestFactory = $itemRequestFactory;
+        $this->sourceSelectionService = $sourceSelectionService;
+        $this->getDefaultSourceSelectionAlgorithmCode = $getDefaultSourceSelectionAlgorithmCode;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Invoice $invoice */
+        $invoice = $observer->getEvent()->getInvoice();
+
+        if ($invoice->getOrigData('entity_id')) {
+            return;
+        }
+
+        $selectionRequestItems = [];
+        foreach ($invoice->getItems() as $invoiceItem) {
+            if (!$this->isValidItem($invoiceItem)) {
+                continue;
+            }
+
+            $itemSku = $invoiceItem->getSku() ?: $this->getSkusByProductIds->execute(
+                [$invoiceItem->getProductId()]
+            )[$invoiceItem->getProductId()];
+            $qty = $this->castQty($invoiceItem->getOrderItem(), $invoiceItem->getQty());
+
+            $selectionRequestItems[] = $this->itemRequestFactory->create([
+                'sku' => $itemSku,
+                'qty' => $qty,
+            ]);
+        }
+
+        if (empty($selectionRequestItems)) {
+            return;
+        }
+
+        $order = $invoice->getOrder();
+        $websiteId = $order->getStore()->getWebsiteId();
+        $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
+
+        /** @var InventoryRequestInterface $inventoryRequest */
+        $inventoryRequest = $this->inventoryRequestFactory->create([
+            'stockId' => $stockId,
+            'items' => $selectionRequestItems
+        ]);
+        $sourceSelectionResult = $this->sourceSelectionService->execute(
+            $inventoryRequest,
+            $this->getDefaultSourceSelectionAlgorithmCode->execute()
+        );
+
+        $this->deductSources($sourceSelectionResult, $stockId);
+    }
+
+    /**
+     * @param InvoiceItemInterface $invoiceItem
+     * @return bool
+     */
+    private function isValidItem(InvoiceItemInterface $invoiceItem): bool
+    {
+        $orderItem = $invoiceItem->getOrderItem();
+        return in_array(
+            $orderItem->getProductType(),
+            [
+                VirtualType::TYPE_VIRTUAL,
+                DownloadableType::TYPE_DOWNLOADABLE
+            ]
+        );
+    }
+
+    /**
+     * @param SourceSelectionResultInterface $sourceSelectionResult
+     * @param int $stockId
+     *
+     * @throws ValidationException
+     * @throws CouldNotSaveException
+     * @throws InputException
+     */
+    private function deductSources(SourceSelectionResultInterface $sourceSelectionResult, int $stockId)
```

Comment:
It could be extracted to dedicated service/class

Reply:
Since update with current 2.3-develop branch there is already a `\Magento\InventoryShipping\Model\SourceDeduction\SourceDeductionServiceInterface` provided by Sergey Mutaf. I've updated my PR to use it.